examples/mm/stable_diffusion/train_sdxl_draftp.py

```diff
@@ -0,0 +1,269 @@
+
+    def _training_strategy(self) -> NLPDDPStrategy:
+        """
+        Returns a DDP or a FSDP strategy passed to Trainer.strategy.  Copied from `sd_xl_train.py`
+        """
+        if self.cfg.model.get("fsdp", False):
+            logging.info("FSDP.")
+            assert (
+                not self.cfg.model.optim.get("name") == "distributed_fused_adam"
+            ), "Distributed optimizer cannot be used with FSDP."
+            if self.cfg.model.get("megatron_amp_O2", False):
+                logging.info("Torch FSDP is not compatible with O2 precision recipe. Setting O2 `False`.")
+                self.cfg.model.megatron_amp_O2 = False
+
+            # Check if its a full-finetuning or PEFT
+            return NLPFSDPStrategy(
+                limit_all_gathers=self.cfg.model.get("fsdp_limit_all_gathers", True),
+                sharding_strategy=self.cfg.model.get("fsdp_sharding_strategy", "full"),
+                cpu_offload=self.cfg.model.get("fsdp_cpu_offload", False),  # offload on is not supported
+                grad_reduce_dtype=self.cfg.model.get("fsdp_grad_reduce_dtype", 32),
+                precision=self.cfg.trainer.precision,
+                ## nn Sequential is supposed to capture the `t_embed`, `label_emb`, `out` layers in the unet
+                extra_fsdp_wrap_module={
+                    UNetModel,
+                    TimestepEmbedSequential,
+                    Decoder,
+                    ResnetBlock,
+                    AttnBlock,
+                    nn.Sequential,
+                    MegatronCLIPRewardModel,
+                    FrozenOpenCLIPEmbedder,
+                    FrozenOpenCLIPEmbedder2,
+                    FrozenCLIPEmbedder,
+                    ParallelLinearAdapter,
+                },
+                use_orig_params=False,  # self.cfg.model.inductor,
+                set_buffer_dtype=self.cfg.get("fsdp_set_buffer_dtype", None),
+            )
+
+        return NLPDDPStrategy(
+            no_ddp_communication_hook=(not self.cfg.model.get("ddp_overlap")),
+            gradient_as_bucket_view=self.cfg.model.gradient_as_bucket_view,
+            find_unused_parameters=False,
+        )
+
+
+def resolve_and_create_trainer(cfg, pop_trainer_key):
+    """resolve the cfg, remove the key before constructing the PTL trainer
+        and then restore it after
+    """
+    OmegaConf.resolve(cfg)
+    with temp_pop_from_config(cfg.trainer, pop_trainer_key):
+        return MegatronStableDiffusionTrainerBuilder(cfg).create_trainer()
+
+
+@hydra_runner(config_path="conf", config_name="draftp_sdxl")
+def main(cfg) -> None:
+
+    logging.info("\n\n************** Experiment configuration ***********")
+    logging.info(f"\n{OmegaConf.to_yaml(cfg)}")
+
+    # set cuda device for each process
+    local_rank = int(os.environ.get("LOCAL_RANK", 0))
+    torch.cuda.set_device(local_rank)
+
+    # TODO: has to be set true for PyTorch 1.12 and later.
```

Comment:
Are we testing on pytorch that old? If not can we remove comment?

But if we are, then we should should guard this (w/ packaging.verion.Version):
```suggestion
    if Version(torch.__version__) > Version('1.12'):
        torch.backends.cuda.matmul.allow_tf32 = True
```

Reply:
not really. I've tested everything on torch2.3.0. I'ts a comment I picked up from `train_sd_draftp.py` -- adding guard